src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+        floatGrids_[floatGridName].get()[i] <= endRange) {
+      boolGrids_[boolGridName].get()[i] = true;
+      num_filled++;
+    }
+  }
+  return num_filled;
+}
+
+void VoxelGrid::fillVoxelSetFromBoolGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                         const std::string& boolGridName,
+                                         bool (*func)(bool)) {
+  assert(boolGrids_.find(boolGridName) != boolGrids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(boolGrids_[boolGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromIntGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                        const std::string& intGridName,
+                                        bool (*func)(int)) {
+  assert(intGrids_.find(intGridName) != intGrids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(intGrids_[intGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromFloatGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                          const std::string& floatGridName,
+                                          bool (*func)(float)) {
+  assert(floatGrids_.find(floatGridName) != floatGrids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if ((*func)(floatGrids_[floatGridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromVector3Grid(std::vector<Mn::Vector3i>& voxelSet,
+                                            const std::string& vector3GridName,
+                                            bool (*func)(Mn::Vector3)) {
+  assert(vector3Grids_.find(vector3GridName) != vector3Grids_.end());
+  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                 m_voxelGridDimensions[2];
+  for (int i = 0; i < gridSize; i++) {
+    if (func(vector3Grids_[vector3GridName].get()[i])) {
+      voxelSet.push_back(reverseHash(i));
+    }
+  }
+}
+
+// 6D SDF - labels each cell as interior (-inf), exterior (+inf), or boundary
+// (0)
+void VoxelGrid::generateInteriorExteriorVoxelGrid() {
+  // create 6 bool grids
+  addBoolGrid("negXShadow");
+  addBoolGrid("posXShadow");
+  addBoolGrid("negYShadow");
+  addBoolGrid("posYShadow");
+  addBoolGrid("negZShadow");
+  addBoolGrid("posZShadow");
+  // fill each grids with ray cast
+  bool hit = false;
+  int ind = 0;
+  std::string gridName = "InteriorExterior";
+  // X axis ray casts
+  for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
```

Comment:
Really should abstract away these raycasts into a little function. A lot of repeated code here where only the axises and string vary.